nova/objects/instance.py

```diff
@@ -1426,11 +1427,30 @@ def fill_faults(self):
         :returns: A list of instance uuids for which faults were found.
         """
         uuids = [inst.uuid for inst in self]
-        faults = objects.InstanceFaultList.get_latest_by_instance_uuids(
+        inst_mappings = objects.InstanceMappingList.get_by_instance_uuids(
             self._context, uuids)
+        inst_cell_mappings = {}
+        for im in inst_mappings:
+            cm = im.cell_mapping
+            cell_id = getattr(cm, 'id', None) if cm else None
+            cm_inst = inst_cell_mappings.get(cell_id)
+            if not cm_inst:
+                cm_inst = {
+                    'cell_mapping': cm,
+                    'instance_uuids': []
+                }
+                inst_cell_mappings[cell_id] = cm_inst
+            cm_inst['instance_uuids'].append(im.instance_uuid)
+
         faults_by_uuid = {}
-        for fault in faults:
-            faults_by_uuid[fault.instance_uuid] = fault
+        for cm_inst in inst_cell_mappings.values():
```

Comment:
```
def _key(x):
  return getattr(x, 'id', None) if x else None

inst_mappings = sorted(inst_mappings, key=_key)
for im, uuids in itertools.groupby(inst_mappings, key=_key):
```
Is that the same? If yes, please change it as it's much shorter and still more expressive in what it does.

Reply:
No, it's not the same but I'll rework the code to look in a similar fashion.

Reply:
Done.